src/exchange/exchange.py

```diff
@@ -77,38 +80,17 @@ def add(self, message: Message) -> None:
     def generate(self) -> Message:
         """Generate the next message."""
         self.moderator.rewrite(self)
-
-        num_times_attempted = 0
-        while num_times_attempted < 3:
-            # we will attempt to generate a response with retries a few times.
-            # if we run into an HTTP error, we will pop the last message until the last
-            # message is a user text message, and then try again. we will do this three
-            # times before giving up.
-            # providers that are hosted on your own machine should not throw HTTP errors,
-            # and could instead configure their own behavior by throwing a different type
-            # of error.
-            try:
-                message, usage = self.provider.complete(
-                    self.model,
-                    self.system,
-                    messages=self.messages,
-                    tools=self.tools,
-                )
-                break
-            except HTTPStatusError:
-                if len(self.messages) <= 1:
-                    # we can't pop any messages, so we have to give up
-                    raise Exception(FAILED_TO_GENERATE_MSG)
-                self.pop_last_message()
-                while len(self.messages) > 1 and self.messages[-1].role == "assistant":
-                    # why 1? because we need to keep at least one user message in the exchange
-                    self.pop_last_message()
-                num_times_attempted += 1
-
-        if num_times_attempted >= REMOVE_MESSAGE_RETRY_TIMES:
-            # we failed to generate a response after three attempts
-            raise Exception(FAILED_TO_GENERATE_MSG)
-
+        try:
+            message, usage = self.provider.complete(
+                self.model,
+                self.system,
+                messages=self.messages,
+                tools=self.tools,
+            )
+        except HTTPStatusError as e:
+            # in the future, if models are hosted on machines we may need to add
+            # other types of exceptions here
+            raise FailedToGenerateMessageError(f"Failed to generate message: {e}") from e
```

Comment:
I'd suggest leaving this as the original HTTP Status Error? that has more actionable information when caught downstream or for debugging (like the request etc)

Reply:
Ok, I figured we might want to handle all `generate` errors similarly, but not attached to it.